### PR TITLE
Fix build for MSVC 19.42

### PR DESCRIPTION
**Pull request by AVAtarMod**
I cannot build async_mqtt without this fix. It may not work at old MSVC version, but I could not find information about version of MSVC which adds used options.

### CODE REVIEW

File: lib/CMakeLists.txt [hunk 1]
```diff
@@ -19,4 +19,20 @@ if(ASYNC_MQTT_USE_LOG)
...
 target_compile_definitions(async_mqtt_separate PRIVATE ASYNC_MQTT_SEPARATE_COMPILATION)
 
+macro(fix_msvc_build)
+    message(STATUS "MSVC build fix applied")
+    target_compile_options(async_mqtt_separate PRIVATE "/Zc:preprocessor" "/bigobj")
+endmacro()
+
+get_property(is_multi_config GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+if(is_multi_config)
+    if ($<STREQUAL:$<CXX_COMPILER_ID>,"MSVC">)
```
Comment (redboltz) [inline lib/CMakeLists.txt:29]:
```suggestion
    if (MSVC)
```

async_mqtt uses `MSVC` at other place. So if you can use it here, please use it.
See https://cmake.org/cmake/help/latest/variable/MSVC.html

File: lib/CMakeLists.txt [hunk 2]
```diff
@@ -19,4 +19,20 @@ if(ASYNC_MQTT_USE_LOG)
 endif()
 target_compile_definitions(async_mqtt_separate PRIVATE ASYNC_MQTT_SEPARATE_COMPILATION)
 
+macro(fix_msvc_build)
+    message(STATUS "MSVC build fix applied")
+    target_compile_options(async_mqtt_separate PRIVATE "/Zc:preprocessor" "/bigobj")
+endmacro()
+
+get_property(is_multi_config GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+if(is_multi_config)
```
Comment (redboltz) [inline lib/CMakeLists.txt:28]: It seems that this part is for the different way to get `CMAKE_CXX_COMPILER_ID`.
I am not 100% sure but `MSVC` could be used for both, if it is right, this part of code could be removed.
And the contents of the macro could be moved into `if (MSVC)` part.

**Review by redboltz — CHANGES_REQUESTED** [review]
@AVAtarMod  Thank you for sending the PR.

I set some of compiler options for MSVC on CI.
See
https://github.com/redboltz/async_mqtt/blob/52432837e35a304ebc5479f64689f13e8d0b83d8/.github/workflows/gha.yml#L178

They should be defined to build separate compilation library too.

`/Zc:preprocessor` `/bigobj` have already been added in your PR. Could you add the rest of options?

File: lib/CMakeLists.txt (hunk 1 above)
Comment (AVAtarMod) [reply lib/CMakeLists.txt]: Done

File: lib/CMakeLists.txt (hunk 2 above)
Comment (AVAtarMod) [reply lib/CMakeLists.txt]: Yes, MSVC could be used, but as far as I know, this is legacy method

**Comment by AVAtarMod** [comment]
> @AVAtarMod Thank you for sending the PR.
> 
> I set some of compiler options for MSVC on CI. See
> 
> https://github.com/redboltz/async_mqtt/blob/52432837e35a304ebc5479f64689f13e8d0b83d8/.github/workflows/gha.yml#L178
> 
> They should be defined to build separate compilation library too.
> 
> `/Zc:preprocessor` `/bigobj` have already been added in your PR. Could you add the rest of options?

Why we need `/D_WIN32_WINNT#0x0A00 /DBOOST_THREAD_VERSION#3 /DBOOST_ASIO_NO_DEPRECATED /EHsc` ?

**Comment by redboltz** [comment]
> Why we need `/D_WIN32_WINNT#0x0A00 /DBOOST_THREAD_VERSION#3 /DBOOST_ASIO_NO_DEPRECATED /EHsc` ?


I realized I was mistaken.

I now believe these options should be left to the user's discretion. If users want to enable them, they can pass them as CMake arguments via CMAKE_CXX_FLAGS. You've already chosen the minimal mandatory options.

Please keep the original code unchanged in this regard.

File: lib/CMakeLists.txt (hunk 2 above)
Comment (redboltz) [reply lib/CMakeLists.txt]: Oh really, I will update exsiting `if (MSVC)` code later.

**Comment by redboltz** [comment]
Merged. Thank you for your contribution.

**Pull request merged**